pgjdbc/src/test/java/org/postgresql/test/jdbc3/ParameterMetaDataTest.java

```diff
@@ -129,4 +140,75 @@ public void testTypeChangeWithUnknown() throws SQLException {
     rs.close();
   }
 
+  @Test
+  public void testGetParameterMetadata_cached() throws SQLException, IllegalAccessException {
+    // given
+    PreparedStatement pstmt =
+        con.prepareStatement("SELECT a FROM parametertest WHERE c = ? AND e = ?");
+
+    pstmt.setString(1, "anything");
+    pstmt.setTimestamp(2, new Timestamp(0L));
+
+    int expectedAmountOfExecutions =
+        binaryMode == BinaryMode.FORCE
+            ? 1
+            : Integer.parseInt(PGProperty.PREPARE_THRESHOLD.getDefaultValue());
+
+    for (int i = 0; i < expectedAmountOfExecutions; i++) {
+      pstmt.executeQuery();
+    }
+
+    // and
+    setUpSpyOnQueryExecutor();
+    QueryExecutor spyOnQueryExecutor = con.unwrap(PgConnection.class).getQueryExecutor();
+
+    // when
+    ParameterMetaData firstCall = pstmt.getParameterMetaData();
+    ParameterMetaData secondCall = pstmt.getParameterMetaData();
+    ParameterMetaData thirdCall = pstmt.getParameterMetaData();
+
+    // then
+    Mockito
+        .verify(spyOnQueryExecutor, Mockito.never())
+        .execute(
+            Mockito.any(Query.class),
+            Mockito.any(ParameterList.class),
+            Mockito.any(ResultHandler.class),
+            Mockito.anyInt(),
+            Mockito.anyInt(),
+            Mockito.anyInt()
+        );
```

Comment:
Frankly speaking, I don't understand what this does.
Could you add a comment that clarifies the assertion in the test?

Reply:
Sure, I'll add the explanation. But briefly, this code tests the behavior we're about to introduce. This means that if the statement is already prepared on the server side, then we should not attempt to send anything via `QueryExecutor`. 

I've almost written a bunch of additional logic to account for forced binary mode and regular mode, since in these two cases, we have different expected amounts of interactions with `QueryExecutor`. Hope it makes sense.